File: LeetCode_Algorithm/863(All_Nodes_Distance_K_in_Binary_Tree)/863.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None
from typing import List
class Solution:
    def makeTree(self, lst, idx):
        if idx >= len(lst):
            return None
        root = TreeNode(lst[idx])
        root.left = self.makeTree(lst, 2*idx+1)
        root.right = self.makeTree(lst, 2*idx+2)
        return root

    def distanceK(self, root: TreeNode, target: TreeNode, k: int) -> List[int]:
        # sol1 49ms 71%
        # graph로 만든 뒤에 target과 k를 넣어서 거리를 측정
        graph = {}; self.visited = []; self.ans = []
        if not root:
            return []
        # self.makeGraph(root, graph)
        # self.solve(target, graph, k)
        # return self.ans

        # sol2 33ms 98%
        # annotate parent
        # node에 부모 속성을 줘서 문제 해결
        # self.par(root, None)        
        # from collections import deque
        # queue = deque([(target, k)])
        # visited = set([target])
        # while queue:
        #     node, k = queue.popleft()
        #     if k == 0:
        #         return [node.val]+[node.val for node, cnt in queue if cnt == 0]
        #     for node in [node.left, node.right, node.par]:
        #         if node and node not in visited:
        #             visited.add(node)
        #             queue.append((node, k-1))
        # return [] 
        # 

        # sol3 35ms 86% 14.5MB 11%
        def dfs(node, parent):
            if not node:
                return ;
            node.parent = parent
            dfs(node.left, node)
            dfs(node.right, node)
        dfs(root, None)
        from collections import deque
        queue = deque([target])
        seen = set([target]); dist = 0
        while queue and dist < k:
            l = len(queue)
            for _ in range(l):
                node = queue.popleft()
                for n in [node.left, node.right, node.parent]:
                    if n and n not in seen:
                        queue.append(n)
                        seen.add(n)
            dist += 1
        return [node.val for node in queue]


    def par(self, node, par):
        if node:
            node.par = par
            self.par(node.left, node)
            self.par(node.right, node)

    def makeGraph(self, node, graph):
        if node not in graph:
            graph[node] = []
        if node.left:
            graph[node].append(node.left)
        if node.right:
            graph[node].append(node.right)
        if node.left:
            if node.left not in graph:
                graph[node.left] = [node]
            self.makeGraph(node.left, graph);
        if node.right:
            if node.right not in graph:
                graph[node.right] = [node]
            self.makeGraph(node.right, graph)

    def solve(self, node, graph, cnt):
        if not node or node.val == None or node in self.visited or cnt < 0:
            return ;
        if cnt == 0:
            self.ans.append(node.val)
            return ;
        self.visited.append(node)
        for temp_node in graph[node]:
            self.solve(temp_node, graph, cnt-1)
        return ;
        
root = Solution().makeTree([3,5,1,6,2,0,8,None,None,7,4], 0)
target = root.left
print(Solution().distanceK(root, target, 2))

File: LeetCode_Algorithm/863(All_Nodes_Distance_K_in_Binary_Tree)/863.js
function TreeNode(val) {
  this.val = val;
  this.left = this.right = null;
}
var makeTree = function (lst, idx) {
  if (idx > lst.length - 1) {
    return null;
  }
  let root = new TreeNode(lst[idx]);
  root.left = makeTree(lst, 2 * idx + 1);
  root.right = makeTree(lst, 2 * idx + 2);
  return root;
};
var makeGraph = function (node, graph) {
  // js의 key값은 스트링이다.
  // https://developer.mozilla.org/ko/docs/Web/JavaScript/Guide/Keyed_collection
  // Object는 문자열을 값에 매핑하는 데 사용되고 Map은 모든 값을 가질 수 있다.
  // Map에 node를 매핑하려고 했는데 [Object Object]라는 결과값이 나왔음
  // Object를 Map의 key로 매핑할 경우 [object Object]라는 스트링으로 매핑이 되기 때문에 사용할 수 없음
  // graph의 key 값을 node의 val 값으로 사용
  if (graph[node.val] == undefined) {
    graph[node.val] = [];
  }
  if (node.left && node.left.val) {
    graph[node.val].push(node.left);
    if (graph[node.left.val] == undefined) {
      graph[node.left.val] = [node];
    }
    makeGraph(node.left, graph);
  }
  if (node.right && node.right.val) {
    graph[node.val].push(node.right);
    if (graph[node.right.val] == undefined) {
      graph[node.right.val] = [node];
    }
    makeGraph(node.right, graph);
  }
};
var solve = function (target, graph, k) {
  console.log(target);
  if (target == null || visited.indexOf(target.val) > -1 || k < 0) {
    return;
  }
  if (k == 0) {
    ans.push(target.val);
    return;
  }
  visited.push(target.val);
  for (let idx in graph[target.val]) {
    solve(graph[target.val][idx], graph, k - 1);
  }
};
var par = function (node, par) {
  if (node) {
    node.par = par;
    par(node.left, node);
    par(node.right, node);
  }
};
var distanceK = function (root, target, k) {
  // sol1 190ms 5%
  // if (!root) {
  //   return []
  // }
  // let graph = new Map(); visited = []; ans = []
  // makeGraph(root, graph)
  // solve(target, graph, k)
  // return ans

  // sol2 98ms 62%
  // par 속성을 추가한 뒤 queue로 나눔
  TreeNode.prototype.par = null;
  if (!root) {
    return [];
  }
  makepar(root, null);

  let queue = [[target, k]];
  let visited = new Set().add(target);

  while (queue.length) {
    [node, k] = queue.shift();
    if (k == 0) {
      ans = [];
      if (node) {
        ans.push(node.val);
      }
      for (let [node, k] of queue) {
        if (node && k == 0) {
          ans.push(node.val);
        }
      }
      return ans;
    }
    for (let [temp, cnt] of [
      [node.left, k - 1],
      [node.right, k - 1],
      [node.par, k - 1],
    ]) {
      if (temp && !visited.has(temp)) {
        visited.add(temp);
        queue.push([temp, cnt]);
      }
    }
  }
  return [];
};

let root = makeTree([3, 5, 1, 6, 2, 0, 8, null, null, 7, 4], 0);
let target = root.left;
console.log(distanceK(root, target, 2));
